src/socket.rs

```diff
@@ -84,13 +84,101 @@ impl ::IntoInner for Socket {
+/// A type with the same memory layout as `c::sockaddr`. Used in converting Rust level
+/// SocketAddr* types into their system representation. The benefit of this specific
+/// type over using `c::sockaddr_storage` is that this type is exactly as large as it
+/// needs to be and not a lot larger. And it can be initialized cleaner from Rust.
+#[repr(C)]
+pub(crate) union SocketAddrCRepr {
+    v4: c::sockaddr_in,
+    v6: c::sockaddr_in6,
+}
+
+impl SocketAddrCRepr {
+    pub(crate) fn as_ptr(&self) -> *const c::sockaddr {
+        self as *const _ as *const c::sockaddr
+    }
+}
+
+fn addr2raw(addr: &SocketAddr) -> (SocketAddrCRepr, c::socklen_t) {
     match *addr {
-        SocketAddr::V4(ref a) => {
-            (a as *const _ as *const _, mem::size_of_val(a) as c::socklen_t)
+        SocketAddr::V4(addr) => {
+            // `s_addr` is stored as BE on all machine and the array is in BE order.
+            // So just transmuting the octects to the `u32` representation works.
+            #[cfg(unix)]
+            let sin_addr = c::in_addr {
+                s_addr: unsafe { mem::transmute::<_, u32>(addr.ip().octets()) },
+            };
+            #[cfg(windows)]
+            let sin_addr = unsafe {
+                let mut s_un = mem::zeroed::<c::in_addr_S_un>();
+                *s_un.S_addr_mut() = mem::transmute::<_, u32>(addr.ip().octets());
+                c::IN_ADDR { S_un: s_un }
+            };
+
+            let sockaddr_in = c::sockaddr_in {
+                sin_family: c::AF_INET as c::sa_family_t,
+                sin_port: addr.port().to_be(),
+                sin_addr,
+                sin_zero: [0; 8],
+                #[cfg(any(
+                    target_os = "dragonfly",
+                    target_os = "freebsd",
+                    target_os = "ios",
+                    target_os = "macos",
+                    target_os = "netbsd",
+                    target_os = "openbsd"
+                ))]
+                sin_len: 0,
+            };
+
+            let sockaddr = SocketAddrCRepr { v4: sockaddr_in };
+            (sockaddr, mem::size_of::<c::sockaddr_in>() as c::socklen_t)
         }
-        SocketAddr::V6(ref a) => {
-            (a as *const _ as *const _, mem::size_of_val(a) as c::socklen_t)
+        SocketAddr::V6(addr) => {
+            #[cfg(unix)]
+            let sin6_addr = {
+                let mut sin6_addr = unsafe { mem::zeroed::<c::in6_addr>() };
+                sin6_addr.s6_addr = addr.ip().octets();
+                sin6_addr
+            };
+            #[cfg(windows)]
+            let sin6_addr = unsafe {
+                let mut u = mem::zeroed::<c::in6_addr_u>();
+                *u.Byte_mut() = addr.ip().octets();
+                c::IN6_ADDR { u }
+            };
+            #[cfg(windows)]
+            let u = unsafe {
+                let mut u = mem::zeroed::<c::SOCKADDR_IN6_LH_u>();
+                *u.sin6_scope_id_mut() = addr.scope_id();
+                u
+            };
```

Comment:
What are your thoughts about splitting `addr2raw` into separate `cfg(windows)` and `cfg(unix)` variants, rather than interleaving like this?  It might be a little easier to follow, event though the `unix` version will still include the various specialization bits.

Reply:
Done. I also splitted the v4 and v6 versions since the function was so long. I think this makes sense and puts most of the code at way lower indentation.